### ISSUE TITLE
function sqrt gives incorrect result

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2999]
Imported from SourceForge on 2024-07-07 08:36:16
Created by **[lvch](https://sourceforge.net/u/lvch/)** on 2013-02-19 06:42:54
Original: https://sourceforge.net/p/maxima/bugs/2549

---

~~~~
(%i1) declare(z,complex); diff(z(q),q)^2*conjugate(diff(z(q),q))^2; sqrt(%);
(%o1) done
(%o2) ('diff(z(q),q,1))^2*conjugate('diff(z(q),q,1))^2
(%o3) 1
~~~~
Should be: `('diff(z(q),q,1))*conjugate('diff(z(q),q,1))`

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 08:36:17
Created by **[amue](https://sourceforge.net/u/amue/)** on 2013-04-18 14:13:27
Original: https://sourceforge.net/p/maxima/bugs/2549/#58b9

---

Maybe [Bug #2550][1] is related to this problem. Could reproduce the problem with wxMaxima 12.04.0 which contains Maxima Version 5.28.0-2.
[1]: https://sourceforge.net/p/maxima/bugs/2550/

Tried the same with

Maxima version: 5.19.2
Maxima build date: 8:55 8/31/2009
host type: i686-pc-mingw32
lisp-implementation-type: GNU Common Lisp (GCL)
lisp-implementation-version: GCL 2.6.8



    declare(z,complex); diff(z(q),q)^2 * conjugate(diff(z(q),q))^2; sqrt(%);

 Got:
 
     (%o2) done
     (%o3) ('diff(z(q),q,1))^2*conjugate('diff(z(q),q,1))^2
     (%o4) sqrt(('diff(z(q),q,1))^2*conjugate('diff(z(q),q,1))^2)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 08:36:21
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-05-24 02:22:36
Original: https://sourceforge.net/p/maxima/bugs/2549/#b397

---

- **labels**: function sqrt gives incorrect result --> Lisp Core - Simplification

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 08:36:24
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-05-24 02:22:36
Original: https://sourceforge.net/p/maxima/bugs/2549/#aa7d

---

Observed in current development version. It appears to be a simplification bug, since it happens in a quoted expression:

    '(sqrt(''%o2));
      => 1

Maxima version: "branch_5_30_base_90_g122d48e_dirty"
Maxima build date: "2013-05-22 18:42:44"
Host type: "i686-pc-linux-gnu"
Lisp implementation type: "SBCL"
Lisp implementation version: "1.0.49"

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 08:36:28
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-08-05 00:14:35
Original: https://sourceforge.net/p/maxima/bugs/2549/#71b8

---

- Description has changed:

Diff:

~~~~

--- old
+++ new
@@ -1,6 +1,7 @@
+~~~~
 (%i1) declare(z,complex); diff(z(q),q)^2*conjugate(diff(z(q),q))^2; sqrt(%);
 (%o1) done
 (%o2) ('diff(z(q),q,1))^2*conjugate('diff(z(q),q,1))^2
 (%o3) 1
-
-Should be: ('diff(z(q),q,1))*conjugate('diff(z(q),q,1)) 
+~~~~
+Should be: `('diff(z(q),q,1))*conjugate('diff(z(q),q,1))`

~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 08:36:31
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-08-20 05:23:32
Original: https://sourceforge.net/p/maxima/bugs/2549/#69ee

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 08:36:35
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-08-20 05:23:32
Original: https://sourceforge.net/p/maxima/bugs/2549/#70c8

---

The incorrect result is fixed by commit 03d2af28. Closing this report.

Incidentally it turns out that the presence of `diff` and `declare(z, complex)` isn't necessary. The simplest example of the bug that I found is:

~~~~
(%i1) sqrt(f(u)^2*conjugate(f(u))^2);
(%o1) 1
~~~~

Now the output is:

~~~~
(%i3) sqrt(f(u)^2*conjugate(f(u))^2);
(%o3) sqrt(conjugate(f(u))^2)*abs(f(u))
~~~~

Maxima suppresses `sqrt(f(u)^2)` --> `abs(f(u))` when `f` is declared complex:

~~~~
(%i4) declare(f, complex);
(%o4) done
(%i5) sqrt(f(u)^2*conjugate(f(u))^2);
(%o5) sqrt(f(u)^2*conjugate(f(u))^2)
~~~~

However, Maxima doesn't know that the derivative of a function of a complex variable is also complex -- Maxima believes that `diff(z(q), q)` is real, so we get now:

~~~~
(%o3) ('diff(z(q),q,1))^2*conjugate('diff(z(q),q,1))^2
(%o4) sqrt(conjugate('diff(z(q),q,1))^2)*abs('diff(z(q),q,1))
~~~~

This is a bug, but it's a different bug. I think it has reported elsewhere, so I will try to find it and put this example in it, otherwise I'll create a report for it.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 08:36:38
Created by **[amue](https://sourceforge.net/u/amue/)** on 2017-08-20 13:31:40
Original: https://sourceforge.net/p/maxima/bugs/2549/#70c8/5464

---

Thank you for fixing this bug. You say that after your fix
~~~
sqrt(f(u)^2*conjugate(f(u))^2);
~~~
gives
~~~
sqrt(conjugate(f(u))^2)*abs(f(u))
~~~
I did not expect this result. This result seems to be based on the assumption that `f(u)` is real. In this case `conjugate(f(u))` is the same as `f(u)`. Therefore I would expect the result
~~~
(abs(f(u)))^2
~~~
Does Maxima know that if `f(u)` is real then `f(u) = conjugate(f(u))`?

BTW: I think about experimenting with Maximas Lisp code but I did not find an easy way to set up a convenient development environment. As I am familiar with Eclipse, I would prefer to use some suitable Eclipse plugin (ABCL ?, I fear Clojure/Counterclockwise will not work). My other problem is that I did not find any documentation that describes the structure of the Maxima code. Therefore I fear I have to spend a lot of time before I can do something useful.
